### ISSUE TITLE
Create clean/ca/los_angeles_pd.py

### ISSUE THREAD

**Issue by gguerrei** [biglocalnews/clean-scraper#18]
- **Agency Slug:** ca_los_angeles_pd
- **Module:** clean/ca/los_angeles_pd.py
- **URL:** https://www.lapdonline.org/senate-bill-1421-senate-bill-16-sb-16/

**Comment by gguerrei** [comment]
MAIN VISION FOR LA:

LA has two systems for getting to the next request page:

1. From main index page -> second index page that holds all child pages (links displayed go directly to nextrequest)
2. From main index page -> second index page that holds most current child pages + more index pages divided by year

In both cases, all URLs come structured (displayed directly on the HTML, no need for reconstruction)

Need two distinct approaches to tackle this:

1. For the first one, straightforward: once on the second index page, write a function that gets child pages URLs and stores them in a list. (**THIS PART IS DONE**)
2. For the second one, more convoluted: will likely require running a second get_index_page_urls while at the same time running the get_child_scrape. Will need to vet which third index page URLs have already been scraped. The challenge here is understanding order of operations, and knowing where each subsequent link should be stored.

Brainstorming the second method:

- For the sake of discussion, let's redefine some terms:
-  Main index page - holds all topics covered by the various public records laws (https://www.lapdonline.org/senate-bill-1421-senate-bill-16-sb-16/)
- Second index page - Officer Involved Shooting (OIS) page. Accessible by clicking the link directly from main index page (https://www.lapdonline.org/sb-1421-ois-main-page/)
- Third index page - yearly pages accessible from the second index page (https://www.lapdonline.org/sb-1421-ois-2023/)
- Child pages - links directly to next request. Accessible from second index page for current year, accessible for other years from respective third index pages. (https://lacity.nextrequest.com/documents?folder_filter=F007-23)

Now that we understand the terms, let's move on to the plan of attack. 
The scraping needs to be broken down into two parts:

PART 1: main index page scrape -> second index page scrape

from there, two possible paths:

PART 2A: child page scrape for current year 
PART 2B: third index page scrape -> child page scrape

To achieve that, here are some goals: 

- Figure out how to write the conditional to run one function instead of the other (or figure out a way to make this selection manual without making the experience worse for the user + fits with other scrapers lexicon and workflow)
- Write a checker function for future scrapes that checks if third index pages were already scraped (since they occupy previous years, no new information will be added to the pages)

**Comment by stucka** [comment]
I am working with this one.

**Comment by stucka** [comment]
@gguerrei , just for the record and to close the loop here, I went in a somewhat different approach with the algorithm, just basic spidering without trying to think too hard about how deep the scraper needed to go on which page.

I identified the parts of the page that would contain links. In this case:
```
        content_divs = soup.findAll("div", {"class": "grid-content"})
        content_divs.extend(soup.findAll("div", {"class": "link-box"}))
```

For each of those things, I went through all the links. I built out a set of links to individual case reports (always on a nextrequest dot com domain) and assumed the other pages were an index; if those pages hadn't been scraped before, I added them to a to-be scraped list. After each was scraped, I removed it from the to-be-scraped pile. When I was out of things to scrape, I was done. (recursion)

So, like:
Trying https://www.lapdonline.org/senate-bill-1421-senate-bill-16-sb-16/
Index scraping pass 1: 1 indexes scraped, 0 case URLs found
...
Trying https://www.lapdonline.org/office-of-the-chief-of-police/constitutional-policing/sustained-complaints-of-unreasonable-excessive-force-failure-to-intervene-in-excessive-force/
Index scraping pass 2: 7 indexes scraped, 443 case URLs found
...
Trying https://www.lapdonline.org/sb-1421-ois-2021/
Index scraping complete, after 28 indexes reviewed.
1,065 case URLs found.

There were some obstacles along the way -- a couple bad URLs broke things, I didn't realize some pages contained things in different types of `div` containers, etc.

Hope to have the code complete in due time. There's a temporary notebook in the platform-84 branch that maybe will soon be a regular Python file in platform-84 that might get pulled into the main in time, but you're welcome to mock it. =)
